chore(docs): document Terraform Cloud usage (#1921)
hashicorp/terraform-cdk: website/docs/cdktf/create-and-deploy/terraform-cloud.mdx

[Workspaces](/cloud-docs/workspaces) function like working directories for distinct Terraform configurations. The workspace is associated with a Terraform configuration and its associated state file. It can also contain variables that you can use for credential management. By default, workspaces are created in [Remote Execution Mode](/cloud-docs/run/remote-operations), which enables Terraform Cloud to store Terraform state and run Terraform operations remotely.

When you create a new workspace you can choose between CLI-Mode (for running `cdktf` directly), VCS-Mode (for letting TFC run in CI mode), and API-Mode (for advanced use-cases).

### CLI-driven Workflow

You can configure the Terraform Cloud CLI-driven workflow for both new and existing CDKTF projects.

### Connect New CDKTF Projects

For new projects, run cdktf init without passing the `--local` flag. CDKTF will use your stored Terraform Cloud credentials to create a new workspace. If you have no stored Terraform Cloud credentials, CDKTF will ask you to login.

> After you login, CDKTF creates a new scaffolded project that is configured to use Terraform state with a remote backend. Will TFC automatically st the workspace to remote execution??

Reviewer: ```suggestion
After you log in, CDKTF creates a new scaffolded project that is configured to store Terraform state in the Terraform Cloud workspace. 
```

### Connect Existing CDKTF Projects

For existing projects, do the following:

1. Create a Terraform Cloud workspace with the CLI-driven workflow.

> 2. Add the RemoteBackend to your CDKTF program.

Reviewer: ```suggestion
2. Add the `RemoteBackend` to your CDKTF application.
```

```ts
import { Construct } from "constructs";
import { App, TerraformStack, RemoteBackend } from "cdktf";

class MyStack extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    // Remote Backend - /docs/backends/types/remote.html
    new RemoteBackend(this, {

>       // You can omit this if you are not using a self-hosted Terraform enterprise instance

Reviewer: ```suggestion
      // You can omit this if you are not using a self-hosted Terraform enterprise instance.
```

      hostname: "app.terraform.io",
      organization: "company",

      workspaces: {
        name: "my-app-prod",
      },
    });

    // define resources here
  }
}

const app = new App();
const myStack = new MyStack(app, "my-stack");
// Instead of defining the backend within the stack you can do it also on the outside like this:
// new RemoteBackend(myStack, {
//     hostname: "app.terraform.io",
//     organization: "company",

//     workspaces: {
//     name: "my-app-prod",
//     },
// });

> // You just need to make sure you configure this before the app.synth() call

Reviewer: ```suggestion
// You must configure the RemoteBackend before the app.synth() call
```

app.synth();
```

> 3. Run `cdktf init` so that you can actually log into TFC

Reviewer: ```suggestion
3. Run `cdktf init` and log in to Terraform Cloud.
```

4. Run `cdktf deploy` / `cdktf plan` / `cdktf destroy` like normally. You can also configure secrets, see the example below.

### VCS-driven Workflow

> You can check your synthesized code into version control and use the VCS-driven workflow. [You can find detailed explanations here.](/cdktf/create-and-deploy/deployment-patterns#deploy-with-terraform-and-terraform-cloud)

Reviewer: ```suggestion
Check your synthesized code into version control and connect your Terraform Cloud workspace to that repository with the VCS-driven workflow. Refer to [Deployment Patterns](/cdktf/create-and-deploy/deployment-patterns#deploy-with-terraform-and-terraform-cloud) for more details about how to use CDKTF with the VCS-driven workflow.
```

      nullable: false, // if passing no
      sensitive: true,
      type: "string",
    });

    // define resources here
  }
}
```

Please bear in mind that the return value is an object with various token representations of the value as attributes. You can pass it as a string via `value.stringValue`, as a number via `value.numberValue`, or as a list via `value.listValue`. This means during synthesization there is no real value in the field, only a [Token](/cdktf/concepts/tokens). Tokens can not be used in dynamic checks during the runtime, e.g. `if (value.listValue.length > 42) {` will always return `false` as tokenized lists have a static length of one item.

## Continuous Integration

To run Terraform Cloud in a CI workflow you can either use [Terraform Clouds VCS option directly](/cdktf/create-and-deploy/deployment-patterns#deploy-with-terraform-and-terraform-cloud) or use a [general-purpose CI to trigger the run in Terraform Cloud](/cdktf/create-and-deploy/deployment-patterns#github-actions-ci-and-terraform-cloud). To learn more about when to use what visit our [Deployment Pattern docs](/cdktf/create-and-deploy/deployment-patterns).

## Policy Enforcement

> Terraform Cloud lets you define [Sentinel Policies](/cloud-docs/sentinel) that can be applied across one or multiple Terraform Cloud workspaces to ensure the provisioned infrastructure follows company-wide security standards / best practices. You can use them with CDKTF if Terraform Cloud is configured as a backend.
> 
> You can use [Run Tasks](/cloud-docs/workspaces/settings/run-tasks) to integrate any pre-apply-checks or post-apply-notifications to your CI workflow via a webhook interface.

Reviewer: ```suggestion
-> **Note:** Sentinel policies and run tasks are available as part of the Terraform Cloud **Team & Governance** upgrade package. Refer to [Terraform Cloud pricing](https://www.hashicorp.com/products/terraform/pricing) for details.

You can define [Sentinel policies](/cloud-docs/sentinel) for one or multiple Terraform Cloud workspaces to ensure that your infrastructure follows company-wide security standards and best practices. You can use Sentinel policies with CDKTF if Terraform Cloud is configured as a backend.

You can also use [Run Tasks](/cloud-docs/workspaces/settings/run-tasks) to directly integrate third-party tools and services at certain stages in the Terraform Cloud run lifecycle. Terraform Cloud uses the status response from the third-party tool to determine if a run should proceed. 
```

### CLI-driven Workflow

You can configure the Terraform Cloud [CLI-driven workflow](/cloud-docs/run/cli) for both new and existing CDKTF projects. This workflow lets you run CDKTF CLI commands to deploy your application execute Terraform operations remotely on Terraform Cloud.

### Connect New CDKTF Projects

For new projects, run cdktf init without passing the `--local` flag. CDKTF uses your stored Terraform Cloud credentials to create a new workspace. If you have no stored Terraform Cloud credentials, CDKTF asks you to log in.

After you log in, CDKTF creates a new scaffolded project that is configured to store Terraform state in the Terraform Cloud workspace. 

### Connect Existing CDKTF Projects

For existing projects, do the following:

1. Create a Terraform Cloud workspace with the CLI-driven workflow.
2. Add the `RemoteBackend` to your CDKTF application.

```ts
import { Construct } from "constructs";
import { App, TerraformStack, RemoteBackend } from "cdktf";

class MyStack extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    // Remote Backend - /docs/backends/types/remote.html
    new RemoteBackend(this, {

>       // You can omit this if you are not using a self-hosted Terraform enterprise instance.

Author: yeah, I think it might be helpful, not sure though

// });
// You must configure the RemoteBackend before the app.synth() call
app.synth();
```

3. Run `cdktf init` and log in to Terraform Cloud.

### VCS-driven Workflow

Check your synthesized code into version control and connect your Terraform Cloud workspace to that repository with the VCS-driven workflow. Refer to [Deployment Patterns](/cdktf/create-and-deploy/deployment-patterns#deploy-with-terraform-and-terraform-cloud) for more details about how to use CDKTF with the VCS-driven workflow.

## Managing Variables and Secrets through Terraform Cloud

Terraform Cloud [variables](/cloud-docs/workspaces/variables) let you define the variables and secrets that Terraform uses during remote operations. You can set variables specifically for each workspace or you can create variable sets to reuse the same variables across multiple workspaces. For example, you could define a variable set of provider credentials and automatically apply it to all of the workspaces using that provider. 

To use variables, set your workspace to [Remote Execution Mode](/cloud-docs/workspaces/settings#execution-mode). Terraform can only access workspace variables when executing remotely on Terraform Cloud. 

> Use the `TerraformVariable` construct to declare variables in your CDKTF application. This creates an undefined [Terraform input variable](/language/values/variables). The following example demonstrates how to create a `TerraformVariable` called `my-var`. 
> 
> ```ts
> import { Construct } from "constructs";
> import { App, TerraformStack, TerraformVariable } from "cdktf";
> 
> class MyStack extends TerraformStack {
>   constructor(scope: Construct, name: string) {
>     super(scope, name);
>     // You can configure the variable 'my-var' in Terraform Cloud and use the value throughout your CDKTF application
>     const value = new TerraformVariable(this, "my-var", {
>       default: "The default value to use",
>       description: "",
>       nullable: false, // if passing no
>       sensitive: true,
>       type: "string",
>     });
> 
>     // define resources here
>   }
> }
> ```

Reviewer: Indenting things so that the numbering will work! :-) 

```suggestion
2. Use the `TerraformVariable` construct to declare variables in your CDKTF application. This creates an undefined [Terraform input variable](/language/values/variables). The following example demonstrates how to create a `TerraformVariable` called `my-var`. 

    ```ts
    import { Construct } from "constructs";
    import { App, TerraformStack, TerraformVariable } from "cdktf";

    class MyStack extends TerraformStack {
      constructor(scope: Construct, name: string) {
        super(scope, name);
        // You can define 'my-var' in Terraform Cloud and use the value in your application.
        const value = new TerraformVariable(this, "my-var", {
          default: "The default value to use",
          description: "",
          nullable: false, // if passing no
          sensitive: true,
          type: "string",
        });

        // define resources here
      }
    }
    ```
```

> After you declare a variable in your CDKTF application, you must [add it to your Terraform Cloud workspace](/cloud-docs/workspaces/variables/managing-variables#add-a-variable) so that Terraform can access the value during plan, apply, and destroy operations.

Reviewer: ```suggestion
3. Add variables that you declared in your CDKTF application to your Terraform Cloud workspace. Refer to [Add a Variable](/cloud-docs/workspaces/variables/managing-variables#add-a-variable) for details. This lets Terraform can access the value during operations.
```

### Variable Return Values

> The return value for `TerraformVariable` is an object with various token representations of the value as attributes. You can pass this object as a string with `value.stringValue`, as a number with `value.numberValue`, or as a list with `value.listValue`. 

Reviewer: ```suggestion
The return value for `TerraformVariable` is an object with various representations of the value as attributes. You can pass this object as a string with `value.stringValue`, as a number with `value.numberValue`, or as a list with `value.listValue`. 
```

> This means that there is no actual value in the variable field when CDKTF synthesizes your application. Instead, CDKTF uses a [Token](/cdktf/concepts/tokens), which represents a value that is unknown until Terraform applies your configuration. You cannot use tokens in dynamic checks during the runtime. For example, `if (value.listValue.length > 42) {` always returns `false` because tokenized lists have a static length of one item.

Reviewer: ```suggestion
This means that there is no actual value in the variable field when CDKTF synthesizes your application. Instead, CDKTF uses a [Token](/cdktf/concepts/tokens), which represents a value that is unknown until Terraform applies your configuration. You cannot use tokens in dynamic checks during runtime. For example, `if (value.listValue.length > 42) {` always returns `false` because tokenized lists have a static length of one item.
```

## Continuous Integration

To run Terraform Cloud in a CI workflow, you can either use [Terraform Cloud's VCS-driven workflow](/cdktf/create-and-deploy/deployment-patterns#deploy-with-terraform-and-terraform-cloud) or use a [general-purpose CI to trigger the run in Terraform Cloud](/cdktf/create-and-deploy/deployment-patterns#github-actions-ci-and-terraform-cloud). 

## Policy Enforcement

-> **Note:** Sentinel policies and run tasks are available as part of the Terraform Cloud **Team & Governance** upgrade package. Refer to [Terraform Cloud pricing](https://www.hashicorp.com/products/terraform/pricing) for details.

> You can define [Sentinel policies](/cloud-docs/sentinel) for one or multiple Terraform Cloud workspaces to ensure that your infrastructure follows company-wide security standards and best practices. You can use Sentinel policies with CDKTF if Terraform Cloud is configured as a backend.

Reviewer: ```suggestion
You can define [Sentinel policies](/cloud-docs/sentinel) for one or multiple Terraform Cloud workspaces to ensure that your infrastructure follows company-wide security standards and best practices. You can use Sentinel policies with CDKTF if Terraform Cloud is configured as a [Remote Backend](/concepts/remote-backends).
```
